server/data/data-api/src/main/java/org/apache/james/rrt/api/CyclicMappingException.java

```diff
@@ -0,0 +1,27 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.rrt.api;
+
+public class CyclicMappingException extends RecipientRewriteTableException {
```

Comment:
Cycle prevention is hard, especially for RRTs.

You only detect the easiest cases within this PR. The `A -> B + B -> A cases`. But there is some other way to be creating a RRT infinite loop that you are not preventing:

 - `A -> B + B -> C + C -> A` (adding intermediates) - these could still be detected upon insertion with Graph cycle algorithms
 - Mobilizing other RRT types like aliases, forwards and the like. Here the domain space is infinite, and you would need to check any RRT belonging to that domain.

My above point is just to highlight you provide only a partial security to the user - and that needs to be clear.

Finally, I want to bring in the **API consistency issues**. Why do you perform such checks on domains? And not on Aliases, forwards, and the like?

While I find it cool to be giving ideas on this concern, I'd like to underline that *this is a different issue* that needs additional discussions. And that should at the very least be separated from this PR.